File: backend/app/schemas/import_data.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class ImportPreviewRow(BaseModel):
    row: int
    data: dict[str, str | None]


class ImportPreviewOut(BaseModel):
    filename: str
    columns: list[str]
    total_rows: int
    sample: list[ImportPreviewRow]
    header_valid: bool
    missing_required: list[str]


class ImportBatchOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    source: str
    filename: str
    status: str
    total_rows: int
    success_rows: int
    failed_rows: int
    error_report: str | None = None
    created_at: datetime

File: frontend/src/api/importData.ts
import { useMutation, useQuery } from '@tanstack/react-query'
import { apiClient } from './client'

export interface ImportPreviewRow {
  row: number
  data: Record<string, string | null>
}

export interface ImportPreview {
  filename: string
  columns: string[]
  total_rows: number
  sample: ImportPreviewRow[]
  header_valid: boolean
  missing_required: string[]
}

export interface ImportBatch {
  id: number
  source: string
  filename: string
  status: string
  total_rows: number
  success_rows: number
  failed_rows: number
  error_report: string | null
  created_at: string
}

function formData(file: File): FormData {
  const data = new FormData()
  data.append('file', file)
  return data
}

export function useImportPreview() {
  return useMutation({
    mutationFn: async (file: File) => {
      const { data } = await apiClient.post<ImportPreview>('/imports/preview', formData(file))
      return data
    },
  })
}

export function useRunImport() {
  return useMutation({
    mutationFn: async (file: File) => {
      const { data } = await apiClient.post<ImportBatch>('/imports', formData(file))
      return data
    },
  })
}

export function useImportBatches() {
  return useQuery({
    queryKey: ['imports'],
    queryFn: async () => {
      const { data } = await apiClient.get<ImportBatch[]>('/imports')
      return data
    },
  })
}
